gemstone/src/swapper/mayan/mayan_swift.rs

```diff
@@ -0,0 +1,412 @@
+use crate::{
+    network::{jsonrpc_call, AlienProvider},
+    swapper::{ApprovalData, ApprovalType},
+};
+use alloy_core::{
+    hex::{decode as HexDecode, encode_prefixed, ToHexExt},
+    primitives::{Address, FixedBytes, U256, U8},
+    sol_types::{SolCall, SolValue},
+};
+
+use gem_evm::{
+    address::EthereumAddress,
+    erc20::IERC20,
+    jsonrpc::{BlockParameter, EthereumRpc, TransactionObject},
+    mayan::swift::swift::IMayanSwift,
+};
+use primitives::Chain;
+use std::{str::FromStr, sync::Arc};
+use thiserror::Error;
+
+pub struct MayanSwift {
+    address: String,
+    provider: Arc<dyn AlienProvider>,
+    chain: Chain,
+}
+
+#[derive(Error, Debug)]
+pub enum MayanSwiftError {
+    #[error("Call failed: {msg}")]
+    CallFailed { msg: String },
+
+    #[error("Invalid response: {msg}")]
+    InvalidResponse { msg: String },
+
+    #[error("ABI error: {msg}")]
+    ABIError { msg: String },
+
+    #[error("Invalid amount")]
+    InvalidAmount,
+}
+
+// Parameter structs with native types
+#[derive(Debug, Clone)]
+pub struct OrderParams {
+    pub trader: [u8; 32],
+    pub token_out: [u8; 32],
+    pub min_amount_out: u64,
+    pub gas_drop: u64,
+    pub cancel_fee: u64,
+    pub refund_fee: u64,
+    pub deadline: u64,
+    pub dest_addr: [u8; 32],
+    pub dest_chain_id: u16,
+    pub referrer_addr: [u8; 32],
+    pub referrer_bps: u8,
+    pub auction_mode: u8,
+    pub random: [u8; 32],
+}
+
+impl OrderParams {
+    pub fn to_contract_params(&self) -> IMayanSwift::OrderParams {
+        IMayanSwift::OrderParams {
+            trader: self.trader.into(),
+            tokenOut: self.token_out.into(),
+            minAmountOut: self.min_amount_out,
+            gasDrop: self.gas_drop,
+            cancelFee: self.cancel_fee,
+            refundFee: self.refund_fee,
+            deadline: self.deadline,
+            destAddr: self.dest_addr.into(),
+            destChainId: self.dest_chain_id,
+            referrerAddr: self.referrer_addr.into(),
+            referrerBps: self.referrer_bps,
+            auctionMode: self.auction_mode,
+            random: self.random.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PermitParams {
```

Comment:
maybe we can reuse `gem_evm::erc2612.rs`, change r/s to `[u8; 32]`

Reply:
tbh I don't think we will use permit at all
I keep this structure just to keep the interface similar to what contract expects